### PR TITLE
Add text transformers, update lists parsing, fix text paragraph style

### PR DESCRIPTION
**Pull request by anton-barkov**

### CODE REVIEW

File: CocoaMarkdown/CMAttributedStringRenderer.m [hunk 1]
```diff
@@ -103,10 +133,19 @@ - (void)parser:(CMParser *)parser didEndHeaderWithLevel:(NSInteger)level
 - (void)parserDidStartParagraph:(CMParser *)parser
 {
     if (![self nodeIsInTightMode:parser.currentNode]) {
-        NSMutableParagraphStyle* paragraphStyle = [NSMutableParagraphStyle new];
-        paragraphStyle.paragraphSpacingBefore = 12;
-        
-        [_attributeStack push:CMDefaultAttributeRun(@{NSParagraphStyleAttributeName: paragraphStyle})];
+        NSParagraphStyle *paragraphStyle = _attributes.textAttributes[NSParagraphStyleAttributeName];
+        if (paragraphStyle)
```
Comment (vdudchak) [inline CocoaMarkdown/CMAttributedStringRenderer.m]:
```suggestion
       NSMutableParagraphStyle *updatedParagraphStyle = paragraphStyle ? [paragraphStyle mutableCopy] : [NSMutableParagraphStyle new];
        updatedParagraphStyle.paragraphSpacingBefore = 12;
        [_attributeStack push:CMDefaultAttributeRun(@{NSParagraphStyleAttributeName: updatedParagraphStyle.copy})];
```

File: CocoaMarkdown/CMPlainTextTransformer.h [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+#import <Foundation/Foundation.h>
+#import "CMNode.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+typedef NSString * _Nonnull (^CMPlainTextTransformerBlock)(NSString *string);
+
+@interface CMPlainTextTransformer : NSObject
+@property (nonatomic, assign, readonly) CMPlainTextTransformerBlock block;
```
Comment (vdudchak) [inline CocoaMarkdown/CMPlainTextTransformer.h]: block is object, it should be captured (strong)

File: CocoaMarkdown/CMAttributedStringRenderer.m [hunk 3]
```diff
@@ -230,7 +269,7 @@ - (void)parserDidEndBlockQuote:(CMParser *)parser
 - (void)parser:(CMParser *)parser didStartUnorderedListWithTightness:(BOOL)tight
 {
     [_attributeStack push:CMDefaultAttributeRun([self listAttributesForNode:parser.currentNode])];
-    [self appendString:@"\n"];
+//    [self appendString:@"\n"];
```
Comment (vdudchak) [inline CocoaMarkdown/CMAttributedStringRenderer.m]: Propose to remove it completely, history will now that it was present before )

File: CocoaMarkdown/CMAttributedStringRenderer.m [hunk 4]
```diff
@@ -20,6 +20,7 @@
 #import "Ono.h"
 
 @interface CMAttributedStringRenderer () <CMParserDelegate>
+@property (nonatomic, strong) NSArray<CMPlainTextTransformer *> *plainTextTransformers;
```
Comment (vdudchak) [inline CocoaMarkdown/CMAttributedStringRenderer.m]: Propose to make NSDictionary<CMNodeKey *, NSArray< CMPlainTextTransformer >> plainTextTransformersMap. Then as key we can create CMNodeKey object (nodeType, headerLevel).

**Comment by anton-barkov** [comment]
@vdudchak all fixed, please review again

File: CocoaMarkdown/CMAttributedStringRenderer.m [hunk 5]
```diff
@@ -85,7 +94,27 @@ - (void)parser:(CMParser *)parser foundText:(NSString *)text
     if (element != nil) {
         [element.buffer appendString:text];
     } else {
-        [self appendString:text];
+        CMNodeType currentNodeType = parser.currentNode.parent.type;
+        NSUInteger currentHeaderLevel = parser.currentNode.parent.headerLevel;
+        
+        CMPlainTextTransformerBlock transfomerBlock = nil;
+        for (CMNodeKey *nodeKey in self.plainTextTransformersMap)
```
Comment (vdudchak) [inline CocoaMarkdown/CMAttributedStringRenderer.m]: Why are you iterating over the map? The idea was to avoid iterating? :)

Comment (vdudchak) [reply CocoaMarkdown/CMAttributedStringRenderer.m]:
```suggestion
        CMNodeKey *nodeKey = [[CMNodeKey alloc] initWithNodeType: currentNodeType headerLevel: currentHeaderLevel];
        CMPlainTextTransformerBlock transfomerBlock = self.plainTextTransformersMap[nodeKey];
```

File: CocoaMarkdown/CMNodeKey.h [hunk 6]
```diff
@@ -0,0 +1,24 @@
+//
+//  CMNodeKey.h
+//  CocoaMarkdown
+//
+//  Created by Anton Barkov on 8/13/19.
+//  Copyright © 2019 Indragie Karunaratne. All rights reserved.
```
Comment (esphynox) [inline CocoaMarkdown/CMNodeKey.h:6]: Is this ok that file created by you, but copyrighted to owner of CocoaMarkdown?

**Review by esphynox — APPROVED** [review]

**Review by vdudchak — APPROVED** [review]

File: CocoaMarkdown/CMNodeKey.h (hunk 6 above)
Comment (anton-barkov) [reply CocoaMarkdown/CMNodeKey.h:6]: I'll leave it this way for now, we have similar copyrights in new files in third party frameworks.

Comment (esphynox) [reply CocoaMarkdown/CMNodeKey.h:6]: 👌

**Pull request merged**